src/netlink.rs

```diff
@@ -67,6 +68,10 @@ impl NetDevice {
         let wifi_info = WifiInfo::new(iface.index).await?;
         let ip = ipv4(&mut sock, iface.index).await?;
         let ipv6 = ipv6(&mut sock, iface.index).await?;
+        let nameserver = read_nameservers()
```

Comment:
Why are you converting only the first nameserver to a sting? I would just keep them as `Vec<IpAddr>` and let `NetDevice` user device how to use the information.

You may find it useful in the `net.rs` block to add `use itertools::Itertools;` so that you can use the `join` function on the list of ips mapped to strings.

<details>
<summary>implementation spoilers</summary>

```diff
diff --git a/src/blocks/net.rs b/src/blocks/net.rs
index d7dea95ab..3b41882b8 100644
--- a/src/blocks/net.rs
+++ b/src/blocks/net.rs
@@ -62,6 +62,7 @@
 use super::prelude::*;
 use crate::netlink::NetDevice;
 use crate::util;
+use itertools::Itertools;
 use regex::Regex;
 use std::time::Instant;
 
@@ -160,7 +161,13 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
                     [if let Some(v) = device.frequency()] "frequency" => Value::hertz(v),
                     [if let Some(v) = device.bitrate()] "bitrate" => Value::bits(v),
                     [if let Some(v) = device.signal()] "signal_strength" => Value::percents(v),
-                    [if let Some(v) = device.nameserver] "nameserver" => Value::text(v.to_string()),
+                    [if !device.nameservers.is_empty()] "nameserver" => Value::text(
+                                                                            device
+                                                                                .nameservers
+                                                                                .into_iter()
+                                                                                .map(|s| s.to_string())
+                                                                                .join(" "),
+                                                                        ),
                     "device" => Value::text(device.iface.name),
                 });
 
diff --git a/src/netlink.rs b/src/netlink.rs
index ca80892ff..a5f0d4f43 100644
--- a/src/netlink.rs
+++ b/src/netlink.rs
@@ -27,7 +27,7 @@ pub struct NetDevice {
     pub ipv6: Option<Ipv6Addr>,
     pub icon: &'static str,
     pub tun_wg_ppp: bool,
-    pub nameserver: Option<String>,
+    pub nameservers: Vec<IpAddr>,
 }
 
 #[derive(Debug, Default)]
@@ -68,9 +68,8 @@ impl NetDevice {
         let wifi_info = WifiInfo::new(iface.index).await?;
         let ip = ipv4(&mut sock, iface.index).await?;
         let ipv6 = ipv6(&mut sock, iface.index).await?;
-        let nameserver = read_nameservers()
+        let nameservers = read_nameservers()
             .await
-            .map(|ns| ns.get(0).map(|ip| ip.to_string()))
             .error("Failed to read nameservers")?;
 
         // TODO: use netlink for the these too
@@ -102,7 +101,7 @@ impl NetDevice {
             ipv6,
             icon,
             tun_wg_ppp: tun | wg | ppp,
-            nameserver,
+            nameservers,
         }))
     }
 ```
</details>

Reply:
> Why are you converting only the first nameserver to a sting? I would just keep them as `Vec<IpAddr>` and let `NetDevice` user device how to use the information.
> 
> You may find it useful in the `net.rs` block to add `use itertools::Itertools;` so that you can use the `join` function on the list of ips mapped to strings.
> 
> implementation spoilers
> ```diff
> diff --git a/src/blocks/net.rs b/src/blocks/net.rs
> index d7dea95ab..3b41882b8 100644
> --- a/src/blocks/net.rs
> +++ b/src/blocks/net.rs
> @@ -62,6 +62,7 @@
>  use super::prelude::*;
>  use crate::netlink::NetDevice;
>  use crate::util;
> +use itertools::Itertools;
>  use regex::Regex;
>  use std::time::Instant;
>  
> @@ -160,7 +161,13 @@ pub async fn run(config: &Config, api: &CommonApi) -> Result<()> {
>                      [if let Some(v) = device.frequency()] "frequency" => Value::hertz(v),
>                      [if let Some(v) = device.bitrate()] "bitrate" => Value::bits(v),
>                      [if let Some(v) = device.signal()] "signal_strength" => Value::percents(v),
> -                    [if let Some(v) = device.nameserver] "nameserver" => Value::text(v.to_string()),
> +                    [if !device.nameservers.is_empty()] "nameserver" => Value::text(
> +                                                                            device
> +                                                                                .nameservers
> +                                                                                .into_iter()
> +                                                                                .map(|s| s.to_string())
> +                                                                                .join(" "),
> +                                                                        ),
>                      "device" => Value::text(device.iface.name),
>                  });
>  
> diff --git a/src/netlink.rs b/src/netlink.rs
> index ca80892ff..a5f0d4f43 100644
> --- a/src/netlink.rs
> +++ b/src/netlink.rs
> @@ -27,7 +27,7 @@ pub struct NetDevice {
>      pub ipv6: Option<Ipv6Addr>,
>      pub icon: &'static str,
>      pub tun_wg_ppp: bool,
> -    pub nameserver: Option<String>,
> +    pub nameservers: Vec<IpAddr>,
>  }
>  
>  #[derive(Debug, Default)]
> @@ -68,9 +68,8 @@ impl NetDevice {
>          let wifi_info = WifiInfo::new(iface.index).await?;
>          let ip = ipv4(&mut sock, iface.index).await?;
>          let ipv6 = ipv6(&mut sock, iface.index).await?;
> -        let nameserver = read_nameservers()
> +        let nameservers = read_nameservers()
>              .await
> -            .map(|ns| ns.get(0).map(|ip| ip.to_string()))
>              .error("Failed to read nameservers")?;
>  
>          // TODO: use netlink for the these too
> @@ -102,7 +101,7 @@ impl NetDevice {
>              ipv6,
>              icon,
>              tun_wg_ppp: tun | wg | ppp,
> -            nameserver,
> +            nameservers,
>          }))
>      }
> ```

Reply:
> Why are you converting only the first nameserver to a sting? I would just keep them as `Vec<IpAddr>` and let `NetDevice` user device how to use the information.

It was a design choice because I wanted to have all the logic in `netlinks.rs` and not in `net.rs`. Eventhough now I think that storing the vector with all the IP's make more sense, and so will change that.

Reply:
> It was a design choice because I wanted to have all the logic in `netlinks.rs` and not in `net.rs`

Ideally `netlink.rs` should contain only netlink-related code.

In this case I guess it is fine to have `resolv.conf` parsing as well, but I agree with @bim9262 that `NetDevice` should just expose a vector of addresses.

Reply:
Actually no, the list of nameservers is not device-specific, so I don't see why it is a part of `NetDevice` in the first place. Please move `read_nameservers()` to `net.rs`.

Reply:
systemd-resolved does allow different DNS servers to be used, not sure if it that would influence the design or not

Reply:
So should I move everything instead to `net.rs`, or leave it in `netlinks.rs`.

Reply:
Since per-interface dns is a thing, I'm okay with it being the way it is right now.